### ISSUE TITLE
Add `Mouvedia/cb-fetch` project

### ISSUE THREAD

**Issue by Mouvedia** [michaelrambeau/bestofjs#60]
https://github.com/Mouvedia/cb-fetch
tags: HTTP client / AJAX, browser compatibility, data fetching, Html & text parsing

**Comment by michaelrambeau** [comment]
Hello @Mouvedia ,
Thank you for the suggestion.
I have done 2 different tests to check the library.

First directly from runkit website: https://runkit.com/npm/cb-fetch

And then, I checked locally:

```js
const request = require('cb-fetch');
request()
  .get('projects.json')
```

But I got an error in both cases.

```
    if (String.isString(input))
               ^

TypeError: String.isString is not a function
```

So I wonder if the library is "production-ready".

Also I wonder how you can compete with libraries like `isomorphic-fetch` (the one that looks the closest form your project since it works in both browser and node.js environment, and it follows the `fetch` browser API), `axios` or `super-agent`...

Anyway, good luck for the fixes!

**Comment by Mouvedia** [comment]
You need to add the dependencies listed in the [`package.json`](https://github.com/Mouvedia/cb-fetch/blob/master/package.json) which has [String.isString](https://www.npmjs.com/package/@string/isstring). Something like adding `require("@string/isstring");`—before the main require—should do.

**But** thanks to your tool Iv found a bug unrelated to your testing: it seems like node.js doesn't have `self` so Ill investigate that one and eventually resubmit.

> Also I wonder how you can compete with…

The specificities of the libraries are its ease of use (e.g. fluent interface—like super-agent) and cross-browser fixes (that are numerous and for the most part missing from the "competitors" alternatives). There are of course more features listed in the README if you are curious.

**Comment by Mouvedia** [comment]
After reviewing the `self` problem I was talking about, it became evident that I was trying to fix a non-issue: node.js doesn't support `fetch` nor `XMLHttpRequest` (or `XDomainRequest`).

My module target is the browser and that's its intended use: you shouldn't test it in a node.js environment. It should be `require` friendly though. Nowadays most browser modules use npm instead of bower, maybe you expected a node module seeing that I was using its packager. Should I emphasize that aspect? It seemed self-evident to me.

@michaelrambeau Apart from the instructions listed [here](https://github.com/Mouvedia/cb-fetch/#importation), do you need help to use it in the browser? If that's still unclear, Id gladly accept your feedbacks so that I can improve the README.

Once you have added the dependencies, here's a simple example that you can run in your browser of choice:
```
request({
  url:          'https://httpbin.org/xml',
  mode:         'cors',
  responseType: 'document'
}).done(function(response) {
	console.log(response);
})
```

**Comment by michaelrambeau** [comment]
Hello @Mouvedia ,
Sorry I missed the fact that it was available only in the browser, not in the node.js, maybe it should be mentioned in the README.

So I am doing some tests, using Browserify to package my JavaScript and I still have the problem about the missing dependency.

I may I have missed something else but when I use a package from npm (using `const request = require('cb-fetch')` for example), I'am not supposed to install manually some external dependencies, or to add extra `require`, I just install the package running `npm install cb-fetch`, that's all. Am I wrong? I mean that's the way all bundlers like Webpack, Browserify or Rollup work, I think.

By the way about bundlers, I'm not sure people still use `YUI` in 2017, maybe the README should focus on instructions about modern tools (unless it's really a feature of the library :)).

About what makes your library specific, since you target browser compatibility, you should mention which browsers are supported (look at https://github.com/github/fetch#browser-support for example). 

You could create tests to show people and add testing report in the README. To run tests in multiple environments, you could use a tool like https://www.browserstack.com/ for example.

For example, look at `superagent`...

![image](https://cloud.githubusercontent.com/assets/5546996/22847193/c1faba90-f02e-11e6-9d17-417ba5000fc5.png)

Or `axios`

![image](https://cloud.githubusercontent.com/assets/5546996/22847217/e743136a-f02e-11e6-8ca0-56be30787603.png)

An other thing, since your library is called `cb-fetch` people expect the library to follow the official `fetch` API, as https://github.com/github/fetch does. So first of all, I think it should return a `Promise`.

If you want your library to stand up from its competitors, highlight what makes the **difference**, show facts (benchmarking ? results of automated tests ?) and add a section in your readme to explain why you should use `cb-fetch` instead of `https://github.com/github/fetch` for example.

That's the only way you can win that competition 😉 

Sorry to help to ask you so many things, I hope it helps you.

**Comment by Mouvedia** [comment]
> …you should mention which browsers are supported…

It's not multi-browser, it's cross-browser. But I understand what you meant by that; I guess I could mention clearly what are the browsers capable of doing async HTTP requests.

> So first of all, I think it should return a Promise.

That would go against the declared purpose: "A truly cross-browser and forward-compatible library to do asynchronous HTTP requests that follows the **callback pattern**."
cb stands for callback and cross-browser.

> Sorry to help to ask you so many things, I hope it helps you.

I wanna thank you for your feedback. Most of your demands are reasonable. Ill take them into account for the next versions.
It seems that having public tests is—understandably—a requirement to be accepted; so Ill withdraw my request for now.

**Comment by michaelrambeau** [comment]
OK @Mouvedia I see, since `Promise` are not available everywhere, you chose to stick to the callback pattern, it makes sense.
So maybe you could emphasise that point in the repo description.

There is no real requirement to be part of bestof.js.org, that's why I feel bad about declining suggestions, I know how much time and energy it takes to create open source projects, so I respect a lot creators like you.

Thank you for your work, once your make a few improvements (the main thing is to make it work with just `npm install` and to explain clearly its purpose in the README.md) I will be very happy to promote your library, OK ?

**Comment by Mouvedia** [comment]
> the main thing is…

It should install the dependencies just fine. But like you said Ill probably have to cater for popular bundlers in an appropriate manner.

> I will be very happy to promote your library

Thanks, Ill reopen the issue once the 1.0 is reached after all it's still just a release candidate.

**Comment by michaelrambeau** [comment]
Thank you for your understanding, keep up the great work!

**Comment by Mouvedia** [comment]
@michaelrambeau I have finally released the 1.0.0!

Here's a playground for you to try it out: https://stackblitz.com/edit/js-nncmxs

I gotta explain the _Html & text parsing_ tag [up there](https://github.com/michaelrambeau/bestofjs/issues/60#issue-206505873): the library normalize the response which means that the handling of XML documents (which is not supported by fetch) is taken care of by the library. Ill let you decide if that warrants that tag or not.

Concerning the browser support, it's cross browser which roughly translate to all browsers since IE5 era.
tl;dr: it's not multi-browser like other libraries are

**Comment by michaelrambeau** [comment]
Hello @Mouvedia 
Thank you for the update!
I wonder why the `package.json` file contains `json3` and `base64` dependencies, because it seems that the source code does not use these dependencies?

**Comment by Mouvedia** [comment]
> I wonder why the package.json file contains json3 and base64 dependencies, because it seems that the source code does not use these dependencies?

Actually it does: those are polyfills for `JSON` and `btoa`. Recent browsers have those but this is geared towards all browsers.

**Comment by michaelrambeau** [comment]
Hello @Mouvedia 
It's online, check here: https://bestof.js.org/tags/http

**Comment by Mouvedia** [comment]
@michaelrambeau awesome, thanks! I am working on the next feature right now.

**Comment by Mouvedia** [comment]
@michaelrambeau cb-fetch is not listed anymore on https://bestof.js.org/tags/http/
Is something amiss?

**Comment by michaelrambeau** [comment]
Hello @Mouvedia 
Thank you for your efforts, I've seen you put a lot of efforts to improve `cb-fetch`.
Currently, to limit the number of projects, only projects with more than 50 stars are shown on the web site.
We will be working to fix this limit, to make the architecture more flexible so that more projects (including yours) can be included.
Sorry about that, stay tuned!
